### ISSUE TITLE
Candid: Spec should not allow for inline types

### ISSUE THREAD

**Issue by hansl** [dfinity/motoko#1686]
The current languages that support Candid generation (Motoko and JavaScript) allow for inline types (defining inlined structures and arrays), but new languages that we want to add support to use Candid (C++, Rust, Java) do not. We also introduce hand written Candid files right now, so users might be tempted to use inlined Candid if we don't explicitly tell them not to.

Multiple solutions;

1. we create temporary names for those structures that users are forced to use (e.g. `struct _FnnameArg0` for the first argument of the `Fnname` function). This is very inconvenient for users and not stable, so users would have to guess which names work.
2. we change the spec to disallow inline non-atomic types. Meaning that all hand 

Inline types also cannot be named in certain contexts (as arguments to functions and values in arrays cannot be named) so it hides the semantics of these types AND it hides type reusability. We don't know if a function arguments matches a function returns semantically (we know syntactically but it's irrelevant for a developer).

To greatly minimize the pain long term as we add more languages, I suggest doing the second option. This does not break our current tooling as Candid generation never use inline types right now. It also is backward compatible as no JS/Motoko generation use inline types (as we don't generate it). So there is minimal impact (ie. no impact) for current code.

There is a concern for third party libraries, but IMO it's very limited and can be worked on case by case basis. The 2 examples I know of are Kotlin and Dart, both of which are done by 55 Foundry and can be worked with easily (and they're also private so no impact on developers outside of those).

TL;DR: Inline types hide semantics, break the easy-to-use nature of Candid and make the experience harder for developers of language like Rust and C++. They could make things even harder in the longer term. They bring little to no value. I suggest simply disallowing them at the language level.
